pkg/aggregator/statsd.go

```diff
@@ -0,0 +1,224 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2025-present Datadog, Inc.
+
+package aggregator
+
+import (
+	"os"
+	"time"
+
+	ddgostatsd "github.com/DataDog/datadog-go/v5/statsd"
+
+	"github.com/DataDog/datadog-agent/comp/core/tagger/origindetection"
+	"github.com/DataDog/datadog-agent/pkg/metrics"
+	"github.com/DataDog/datadog-agent/pkg/metrics/event"
+	"github.com/DataDog/datadog-agent/pkg/metrics/servicecheck"
+	taggertypes "github.com/DataDog/datadog-agent/pkg/tagger/types"
+)
+
+// NewStatsdDirect creates a direct interface to the dogstatsd demultiplexer, but exposing the statsd.ClientInterface
+func NewStatsdDirect(demux DemultiplexerWithAggregator) (ddgostatsd.ClientInterface, error) {
+	eventsChan, serviceCheckChan := demux.GetEventsAndServiceChecksChannels()
+	return &statsdDirect{
+		demux:            demux,
+		eventsChan:       eventsChan,
+		serviceCheckChan: serviceCheckChan,
+		origin: taggertypes.OriginInfo{
+			LocalData:     origindetection.LocalData{ProcessID: uint32(os.Getpid())},
+			ProductOrigin: origindetection.ProductOriginDogStatsD,
+		},
+	}, nil
+}
+
+var _ ddgostatsd.ClientInterface = (*statsdDirect)(nil)
+
+type statsdDirect struct {
+	demux            DemultiplexerWithAggregator
+	eventsChan       chan []*event.Event
+	serviceCheckChan chan []*servicecheck.ServiceCheck
+	origin           taggertypes.OriginInfo
+}
+
+func (s statsdDirect) Gauge(name string, value float64, tags []string, rate float64) error {
+	s.demux.AggregateSample(metrics.MetricSample{
+		Name:       name,
+		Value:      value,
+		Mtype:      metrics.GaugeType,
+		Tags:       tags,
+		SampleRate: rate,
+		Timestamp:  float64(time.Now().Unix()),
+		OriginInfo: s.origin,
+	})
+	return nil
+}
+
+func (s statsdDirect) GaugeWithTimestamp(name string, value float64, tags []string, rate float64, timestamp time.Time) error {
+	s.demux.AggregateSample(metrics.MetricSample{
+		Name:       name,
+		Value:      value,
+		Mtype:      metrics.GaugeWithTimestampType,
```

Comment:
So I'm not gonna lie: I don't actually know why we have the `FooWithTimestampType` constants, but for timestamped metrics explicitly, we actually want to send them through the no aggregation pipeline.

Two things to do:

- use the `WithTimestamp`-less version of the metric type (so `metrics.GaugeWithTimestampType` -> `metrics.GaugeType`, same for counts)
- call `s.demux.SendSamplesWithoutAggregation` instead (which takes a slice of `metrics.MetricSample`, so you'll just have to encapsulate)